File: port/lp-solver.js
/**
 * LP solver abstraction layer.
 * 
 * This module provides an abstraction for solving LP problems from text format.
 * Uses the HiGHS solver via highs-js.
 */

/**
 * Result from LP solver.
 */
export class SolverResult {
    /**
     * @param {string} status - optimization status
     * @param {Object<string, number>} variable_values - dict mapping variable name to solution value
     */
    constructor(status, variable_values) {
        this.status = status;
        this.variable_values = variable_values;
    }
    
    /**
     * Check if optimization succeeded.
     * @returns {boolean}
     */
    is_optimal() {
        return this.status === 'Optimal';
    }
}

// Singleton HiGHS instance
let _highs_instance = null;

/**
 * Initialize HiGHS solver (loads the WASM module).
 * Should be called once before using solve_lp.
 * 
 * @returns {Promise<Object>} HiGHS instance
 */
export async function init_highs() {
    if (_highs_instance !== null) {
        return _highs_instance;
    }
    
    // Load HiGHS module from CDN
    _highs_instance = await Module({
        locateFile: (file) => `https://lovasoa.github.io/highs-js/${file}`
    });
    
    return _highs_instance;
}

/**
 * Solve LP problem from CPLEX LP format text.
 * Automatically initializes HiGHS if not already initialized.
 * 
 * Precondition:
 *     lp_text is valid CPLEX LP format
 *     HiGHS library is loaded (via script tag or import)
 * 
 * Postcondition:
 *     returns SolverResult with optimization status and variable values
 * 
 * @param {string} lp_text - LP problem in CPLEX LP format
 * @returns {Promise<SolverResult>} containing status and variable values
 */
export async function solve_lp(lp_text) {
    // Auto-initialize if not already done
    if (_highs_instance === null) {
        await init_highs();
    }
    
    // Solve the LP problem using HiGHS
    const solution = _highs_instance.solve(lp_text);
    
    // Extract variable values from solution
    const variable_values = {};
    if (solution.Columns) {
        for (const [var_name, var_info] of Object.entries(solution.Columns)) {
            // Include all variables, even those with zero values
            variable_values[var_name] = var_info.Primal || 0;
        }
    }
    
    return new SolverResult(solution.Status, variable_values);
}

File: lp_solver.py
"""LP solver abstraction layer.

This module provides an abstraction for solving LP problems from text format.
In Python, uses the mip package. When porting to JavaScript, replace with HiGHS.
"""

import os
import tempfile

from mip import Model, OptimizationStatus as MipStatus


# Solver status constants - abstracted from mip
class SolverStatus:
    """Optimization status constants."""
    OPTIMAL = "OPTIMAL"
    INFEASIBLE = "INFEASIBLE"
    UNBOUNDED = "UNBOUNDED"
    FEASIBLE = "FEASIBLE"
    LOADED = "LOADED"
    ERROR = "ERROR"


class SolverResult:
    """Result from LP solver.
    
    Attributes:
        status: optimization status (SolverStatus constant)
        variable_values: dict mapping variable name to solution value
    """
    def __init__(self, status: str, variable_values: dict[str, float]):
        self.status = status
        self.variable_values = variable_values
    
    def is_optimal(self) -> bool:
        """Check if optimization succeeded."""
        return self.status == SolverStatus.OPTIMAL


def solve_lp(lp_text: str) -> SolverResult:
    """Solve LP problem from CPLEX LP format text.
    
    Precondition:
        lp_text is valid CPLEX LP format
    
    Postcondition:
        returns SolverResult with optimization status and variable values
    
    Args:
        lp_text: LP problem in CPLEX LP format
    
    Returns:
        SolverResult containing status and variable values
    """
    # Write LP text to temp file
    with tempfile.NamedTemporaryFile(mode='w', suffix='.lp', delete=False) as f:
        f.write(lp_text)
        temp_lp_path = f.name
    
    try:
        # Load LP file with mip and optimize
        model = Model()
        model.verbose = 0
        model.read(temp_lp_path)
        model.optimize()
        
        # Extract variable values
        variable_values = {}
        for var in model.vars:
            if var.x is not None:
                variable_values[var.name] = var.x
        
        # Map mip status to abstracted status
        status_map = {
            MipStatus.OPTIMAL: SolverStatus.OPTIMAL,
            MipStatus.INFEASIBLE: SolverStatus.INFEASIBLE,
            MipStatus.UNBOUNDED: SolverStatus.UNBOUNDED,
            MipStatus.FEASIBLE: SolverStatus.FEASIBLE,
            MipStatus.LOADED: SolverStatus.LOADED,
        }
        status = status_map.get(model.status, SolverStatus.ERROR)
        
        return SolverResult(status, variable_values)
    
    finally:
        # Clean up temp file
        if os.path.exists(temp_lp_path):
            os.remove(temp_lp_path)

